$$
Белорусский\;государственный\; университет
$$
$$
Механико-математический\;факультет
$$
$$
Кафедра\;дифференциальных\;уравнений\; и\; системного\; анализа
$$
$ $

$$
 \Large\bf Математические\; основы\; защиты\; информации
$$

# Тема 6. Криптосистемы с открытым ключом. Криптосистема RSA

$ $

доцент Чергинец Дмитрий Николаевич

# 1. Введение в криптосистемы

##  Открытый и зашифрованный текст
Пусть $\;A\;$ и $\;B\;$ -- конечные множества, которые будем называть алфавитами. 

Защищаемую информацию,  состоящую из конечного объединения элементов  множества $\;A,\;$ будем называть <font color='blue'>открытым текстом</font> (Plaintext) или сообщением.
 
Преобразованный открытый текст, состоящий из конечного объединения элементов множества $\;B,\;$ называется <font color='blue'>шифрованным текстом</font> или <font color='blue'>шифротекстом</font> (Сiphertext). 


Через $\;X\;$ и $\;Y\;$ обозначим множества открытых текстов и шифрованных текстов соответственно. $\;A\;$ и $\;B\;$ назовём алфавитом открытого текста и шифрованного текста
соответственно.

## Симметричная криптосистема
Функцию $\;E_k:X\rightarrow Y\;$ 
будем называть  <font color='blue'>функцией шифрования</font> (Encrypt).  

Параметр $\;k\;$ функции $\;E_k\;$ называется <font color='blue'>ключом</font> и принадлежит множеству ключей $\;K.\;$



Функция
$\;D_k:Y\rightarrow X\;$  называется <font color='blue'>функцией дешифрования</font> (Decrypt).


<font color='blue'>Шифром</font> или <font color='blue'>криптосистемой</font> называется набор
$$(A,B,X,Y,K,E_k,D_k),
$$
удовлетворяющий требованию 
$$D_k(E_k(x))=x$$ для
каждого $\;x\in X\;$ и $\;k\in K.\;$


## AES
 В 1974 году Национальное бюро стандартов США объявило конкурс    на стандарт шифрования. Данный конкурс выиграл симметричный     шифр DES, разработанный компанией IBM. 

 В 1997 году объявлен конкурс на новый шифр.


В  2002 году на смену DES пришел AES (Advanced Encryption Standard), также известный как Rijndael (Рейндал).


## Парадокс симметричных криптосистем
Криптосистема, в которой для шифрования и дешифрования используется один и тот же (секретный) ключ, называется симметричной.

Для того чтобы два человека могли обменяться секретной  информацией, ее нужно зашифровать при помощи секретного ключа,  но теперь возникает проблема передачи секретного ключа для   дешифрования, в этом заключается парадокс симметричных криптосистем.

## Шифры с открытым ключом
Шифры, в которых для шифрования используется один ключ, а для дешифрования -- другой, называются <font color='blue'>асимметричными</font>  или <font color='blue'>криптосистемами с открытым ключом</font>. 

![PKK](images/PKK.png)

## Шифры с открытым ключом
Криптосистемой с открытым ключом называется система
$$
  (X,Y,(k_e,k_d)\in K,E_{k_e},D_{k_d}),
$$
где алгоритмы шифрования
$$
  E_{k_e}:X\rightarrow Y
$$
и дешифрования
$$
  D_{k_d}:Y\rightarrow X
$$
являются открытыми, шифротекст $\;c\;$ и <font color='blue'>открытый ключ</font> $\;k_e\;$ (public key) могут передаваться по     незащищенному каналу, <font color='blue'>секретный ключ</font> $\;k_d\;$ (private key) является секретным.    

## Требования к криптосистеме с открытым ключом
Основы криптографии с открытым ключом заложены Уитфильдом Диффи и Мартином Хеллманом в статье "Новые направления в  криптографии", опубликованной в 1976 г.

У. Диффи и М. Хеллман сформулировали основные требования, предъявляемые к криптосистемам с открытым ключом:

- Генерация каждой пары ключей $\;(k_e,k_d)\in K\;$ является полиномиальным алгоритмом.

- Для каждой пары ключей $\;(k_e,k_d)\in K\;$ и каждого открытого текста $\;m\in M\;$ функции $\;E\;$ и $\;D\;$ полиномиальны.

- Противник, зная открытый ключ $\;k_e\;$ и шифротекст $\;c,\;$ при попытке вычислить открытый текст $\;m\;$ наталкивается на непреодолимую вычислительную проблему.

# 2. Односторонние функции
 Отображение $\;f:X\rightarrow Y\;$ называется <font color='blue'>односторонней</font> (однонаправленной) <font color='blue'>функцией</font> (one-way function), если 
 
1) существует полиномиальный алгоритм, для каждого $\;x\in X\;$ вычисляющий $f(x);$

2) не существует полиномиального алгоритма, вычисляющего $\;f^{-1}(y)\;$ для каждого $\;y\in f(X).\;$

Если использовать понятие односторонней функции только для криптосистем, то функция $\;f\;$ должна быть инъективной. Но понятие односторонней функции часто распространяют на криптографические функции хеширования, которые как правило не инъективны.

## Односторонние функции с лазейкой
Если для односторонней функции $\;f\;$ существует дополнительная информация (лазейка) $\;k,\;$    зная которую $\;f_k^{-1}(y)\;$ вычисляется за полиномиальное время, то данная функция $\;f\;$ называется <font color='blue'>односторонней функцией с лазейкой</font> (one-way trap-door function).

## Правдоподобный пример односторонней функции
Пусть 
$$\;e,n\in\mathbb{N},\qquad 1<e<\varphi(n),\qquad\gcd(e,\varphi(n))=1,\;
$$
будем также считать, что $\;\varphi(n)\;$ неизвестно и его достаточно трудно вычислить.
      
      
Рассмотрим функцию 
$$
\;f:\mathbb{Z}_n^*\rightarrow\mathbb{Z}_n^*,\;
$$
действующую по правилу
$$
  f(x):=x^e \mod n.
$$

$\;f(x)\;$ вычисляется за полиномиальное время.
      
Пока не найдено полиномиального алгоритма, вычисляющего $\;f^{-1}(y),\;$ т. е. решающего уравнение 
$$
x^e\equiv b \mod  n,
$$ 
но и не доказано, что такого алгоритма не существует.

##  Пример функции с лазейкой
Однако, если мы для функции
$
  \;f(x):=x^e \mod n\;
$
узнаем число $\;d,\;$ удовлетворяющее условию 
$$
e d \equiv 1\mod\varphi(n) \qquad \Leftrightarrow \qquad
e d = 1 + z\varphi(n),\quad z\in\mathbb{Z},
$$
то обратную функцию можно определить формулой
$$
  f^{-1}(y):=y^d\mod n.
$$
Действительно,
$$
y^d\equiv x^{e d}\equiv x x^{z\varphi(n)}\equiv x\mod n.
$$
Мы нашли $\;f^{-1},\;$ поэтому $\;f\;$ --- инъекция, а так как она определена на конечном множестве $\;\mathbb{Z}_n^*,\;$ то $\;f\;$ --- биекция. 

Таким образом, если считать, что не существует полиномиального алгоритма, решающего уравнение
$$
x^e\equiv    b\mod n,
$$
то функция $\;f\;$ является односторонней функцией с лазейкой.

##  Классы P, NP
<font color="blue"> Классом P</font> называют множество задач, для которых существует полиномиальный алгоритм решения.    
		
<font color="blue">Классом NP</font> называют множество задач, 	решение которых можно проверить за полиномиальное время.
		
То есть для задач из класса NP существует полиномиальное доказательство: для каждого предъявленного решения существуют  некие данные, при помощи которых за полиномиальное время можно доказать, что предъявленное решение действительно является решением.  

## Задача факторизации чисел принадлежит NP
Задача разложения числа $\;n\;$ на множители принадлежит классу NP, так как предъявленное решение задачи
$$
p_1,p_2,\dots, p_k
$$
можно проверить за полиномиальное время, перемножив предъявленные числа и сравнив результат с $\;n:\;$
$$
p_1* p_2*\dots* p_k == n?
$$

Дополнительных данных для доказательства в данном случае не требуется.    

## P=NP?
		
-  Очевидно, что $P\subset NP.$ 
		
- Верно ли обратное включение $NP\subset P$ ? 
		
- Данная задача является одной из семи Задач тысячелетия.    
		
- Проблему $\;P = NP\;$ можно переформулировать таким образом: если 	положительный ответ на какой-то вопрос можно быстро проверить (за 		полиномиальное время), то правда ли, что ответ на этот вопрос 		можно быстро найти (за полиномиальное время и используя 		полиномиальную память)?

## Существование односторонних функций
На данный момент не доказано существование ни одной односторонней функции. 

**Теорема.**  
Если $\;P=NP,\;$ то односторонних функций не существует.

**Доказательство.**  
Действительно, для $\;y\in Y\;$ существует полиномиальный алгоритм проверки того, что $$\;f(x)=y.\;
$$ 
   
Если $\;P=NP,\;$ то должен существовать и полиномиальный алгоритм, вычисляющий 
$$
\;f^{-1}(y).\;
$$


# 3. Криптосистема RSA
Название криптосистемы RSA образовано заглавными буквами фамилий ее авторов:  Рональда Ривеста, Ади Шамира и Леонарда Адлемана.
		
В августе 1977 года в колонке "Математические игры" Мартина Гарднера в журнале Scientific American появилось первое описание криптосистемы RSA. 
		
Читателям также было предложено дешифровать фразу. За расшифровку была обещана награда в 100 долларов США. Фразу расшифровали лишь в 1995 году.

## Генерация ключей в RSA
Input:  $\;L\in\mathbb{N}\;$ -- длина простых чисел.  
  
Output: $\;k_e\;$ -- открытый ключ,
		
$\qquad\;\;$   $\;k_d\;$ -- секретный  ключ.
		
		
1. Выбираем два случайных простых числа $\;p, q,\;$ состоящие из $\;L\;$ бит.
			
2. Вычисляем $\;n:=p q,\;$ $\;\varphi(n):=(p-1)(q-1).\;$
			
3. Случайным образом выбираем натуральное число $\;e,\;$ удовлетворяющее условиям
$$
1<e<\varphi(n),\qquad \gcd(e,\varphi(n))=1.
$$
4. При помощи расширенного алгоритма Евклида вычисляем 
$$
d:=e^{-1}\mod\varphi(n)
$$
or
$$
ed \equiv 1 \mod\varphi(n)
$$
5. Выдаем результат: $\;k_e:=(e,n),\;$ $\;k_d:=(d,n).$
			

## <font color='red'>Задание 1.</font>
Используя функции из предыдущих лабораторных занятий, реализовать алгоритм генерации ключей криптосистемы RSA. 

In [1]:
from random import randint
import math

def st(n):
    s = 0
    t = n - 1
    while t % 2 == 0:
        s += 1
        t = t // 2
    return s, t

def WitnessQ(a, n, s = 0, t = 0):  
    if n % 2 == 0:
        return False
    if s == 0:
        s, t = st(n)    
    if math.gcd(a,n) > 1:
        return False
    b = pow(a,t,n)
    if b == 1 or b == n - 1:  # модуль - число положительное
        return True
    for _ in range(1,s):
        b = b**2 % n
        if b == n - 1:
            return True
    return False

def MillerRabinTest(n, k = 10):
    if n % 2 == 0:
        return False
    s,t = st(n)
    for _ in range(k):
        a = randint(2,n - 1)
        if not WitnessQ(a,n,s,t):
            return False
    return True

def generates_prime_number(b):
    while True:
        n = randint(2**(b - 1),2**b - 1)
        if MillerRabinTest(n):
            return n
        else:
            continue
            
def ExtendedGCD(a, b):  # расширенный алгоритм Евклида
    m11, m12 = 1, 0
    m21, m22 = 0, 1
    while b:
        q = a // b
        r = a % b
        a, b = b, r
        m11, m12 = m12, m11 - m12*q
        m21, m22 = m22, m21 - m22*q 
    return (m11,m21)

def RSA_key_generation(L: int) -> dict:
    
    """
    L is the length of the primes
    
    k_e is the open key
    k_d is the close key
    """
    
    p, q = generates_prime_number(L), generates_prime_number(L)
    n = p * q
    phi = (p - 1) * (q - 1)
    
    while True:
        e = randint(1 + 1, phi - 1)  # чтобы обеспечить выполнение неравенства
        
        if math.gcd(e, phi) == 1:
            break
        else:
            continue
            
    d, _ = ExtendedGCD(e, phi)  # обосновать через коэффициенты Безу
    
    return {'k_e': (n, e), 'k_d': (n, d)}

In [2]:
RSA_key_generation(1024)

{'k_e': (19302030588325833401116640268375889199660219298866854740577481208089230565816899071195205938773097600529248349210682640297306744088714446045904223737218609723275529197883754833737831559553155284991977837647312859809077844939077692938454097920138115317470277276884003973797458824807919354356254044562210517521442347585009884412038704246363740855791519238388522995890155952283213289375946138123892429222391456837714352282888545314234422552551550811541924528867190043277732122244111761773142347302506298723093882311060822099905115626230646667002404090970539231782488758537598391602702824189397039707591718096927358609637,
  1299570299298948342564613835255516871805297950035107444993440685453468810122888077284802739540107579105141915446485451687537624633572846981082007005443434980152823456192562582485253604016925813147887133455757554326348607463616034600917336082501931116375733900077116622970750871680114287281944417549727193323540897080269266476601090324268689160496254794956169303660946603

## <font color='red'>Задание 2.</font>
Почему на практике в качестве публичной экспоненты $\;e\;$ часто выбирают пятое число Ферма:
$$
  F_4=2^{2^4} + 1=65537.
$$  

Обычно в качестве $e$ берут простые числа, содержащие небольшое количество единичных бит в двоичной записи, например,
простые из чисел Ферма: 17, 257 или 65537, так как в этом случае время, необходимое для шифрования с использованием
быстрого возведения в степень, будет меньше (прочитать можно <a href = "https://ru.wikipedia.org/wiki/RSA#%D0%90%D0%BB%D0%B3%D0%BE%D1%80%D0%B8%D1%82%D0%BC_%D1%81%D0%BE%D0%B7%D0%B4%D0%B0%D0%BD%D0%B8%D1%8F_%D0%BE%D1%82%D0%BA%D1%80%D1%8B%D1%82%D0%BE%D0%B3%D0%BE_%D0%B8_%D1%81%D0%B5%D0%BA%D1%80%D0%B5%D1%82%D0%BD%D0%BE%D0%B3%D0%BE_%D0%BA%D0%BB%D1%8E%D1%87%D0%B5%D0%B9:~:text=%D0%BE%D0%B1%D1%8B%D1%87%D0%BD%D0%BE%20%D0%B2%20%D0%BA%D0%B0%D1%87%D0%B5%D1%81%D1%82%D0%B2%D0%B5,%2C%20%D0%B1%D1%83%D0%B4%D0%B5%D1%82%20%D0%BC%D0%B5%D0%BD%D1%8C%D1%88%D0%B5%3B">здесь<a>).

## Шифрование и дешифрование в RSA
**Шифрование.**  
Функция шифрования 	
$$
\;E_{k_e}:\mathbb{Z}_n\rightarrow\mathbb{Z}_n\;
$$
определяется правилом
$$
 E_{k_e}(m):=m^e\mod n.
$$

**Дешифрование.**  
Функция дешифрования 	
$$
  D_{k_d}:\mathbb{Z}_n\rightarrow\mathbb{Z}_n
$$
задается формулой
$$
	D_{k_d}(c):=c^d\mod n.
$$

Отметим, что все три приведенных выше алгоритма являются полиномиальными.

## <font color='red'>Задание 3.</font>
Используя функции из предыдущих лабораторных занятий, реализовать алгоритмы шифрования и дешифрования криптосистемы RSA. 

In [3]:
key_dict = RSA_key_generation(8)
k_e = key_dict['k_e']
k_d = key_dict['k_d']

In [4]:
def encryption_E(m: int, k_e: tuple) -> int:
    
    """ encryption with method RCA """
    
    return pow(m, k_e[1], k_e[0])

In [5]:
def decryption_D(c: int, k_d: tuple) -> int:
    
    """ decryption with method RCA """
    
    return pow(c, k_d[1], k_d[0])

In [6]:
m = 673  # возникают проблемы с отрицательными числами из-за их представления
c = encryption_E(m, k_e)
m_new = decryption_D(c, k_d)

m == m_new

True

## Корректность RSA
**Теорема 1.**  
Пусть $\;n=p q,\;$ $\;p\neq q\;$ -- простые, $\;e\in\mathbb{N},\;$ $\;1<e<\varphi(n),\;$      $\;\gcd(e,\varphi(n))=1.\;$ 
      
Тогда для каждого $\;m\in\mathbb{Z}_n\;$ справедливо равенство
$$
  D_{k_d}(E_{k_e}(m))=m.
$$



## Доказательство корректности

Исходя из определений функций $\;E\;$ и $\;D\;$ нам необходимо доказать, что
$$
   m^{e d}\equiv m \mod n.
$$
Так как $\;d:=e^{-1}\mod \varphi(n),\;$ то найдется такое $\;z\in\mathbb{Z},\;$ что $\;ed= 1+z\varphi(n).\;$


**1$.$ Пусть сначала $\;\gcd(m,n)=1.$**   
    
Тогда по теореме Эйлера
$$
   m^{e d}=m^{1+z\varphi(n)}\equiv m (m^{\varphi(n)})^z\equiv m\mod n.
$$

**2$.$ Пусть $\;\gcd(m,n)=n.\;$**  Тогда 
 $$
 m^{ed}\equiv 0\equiv m\mod n.
 $$


**3$.$ Пусть, наконец, $\;\gcd(m,n)=p.$** 
    
Из того, что $\;\gcd(m,q)=1,\;$ имеем 
$$
  \left\{\begin{array}{ll}m^{e d}\equiv 0\equiv m & \mod p,\\
  m^{e d}\equiv(m)^{1+z\varphi(p)\varphi(q)}\equiv m (m^{\varphi(q)})^{z\varphi(p)}\equiv m & \mod q.
\end{array}\right.
$$
Так как $\;\gcd(p,q)=1,\;$ то по свойству сравнений $\;m^{ed}\equiv m\mod n.$

## Эквивалентность задачи факторизации и вычисления функции Эйлера
**Теорема 2.**  
Задача разложения на множители числа $\;n=p q\;$ и задача вычисления функции Эйлера $\;\varphi(n)\;$ полиномиально эквивалентны.
$$
 n=pq\qquad\Leftrightarrow\qquad\varphi(n)
$$ 
**Доказательство**  $\;\;n=pq \quad\Rightarrow\quad\varphi(n)$  
Если известно разложение числа $\;n\;$ на множители $\;p, q,\;$ то значение функции Эйлера можно вычислить за одну арифметическую операцию 
$$
  \;\varphi(n)=(p-1)(q-1).
$$

$\varphi(n)\quad\Rightarrow\quad n=pq$

Обратно, если известно $\;\varphi(n),\;$ то множители $\;p,q\;$ удовлетворяют системе
$$
  \left\{\begin{array}{l}
    p q = n,\\
    (p-1)(q-1) = \varphi(n);
\end{array}
\right.\Rightarrow
  \left\{\begin{array}{l}
    p q = n,\\
    p+q = n+1-\varphi(n).
\end{array}
\right.
$$
По теореме Виета $\;p,q\;$ -- корни уравнения
$$
  x^2 - (n+1-\varphi(n))x +n = 0.
$$
Таким образом
$$
  p=\frac{n+1-\varphi(n)+\sqrt{(n+1-\varphi(n))^2-4n}}{2},\;  
$$
$$
q=\frac{n+1-\varphi(n)-\sqrt{(n+1-\varphi(n))^2-4n}}{2}.
$$

## <font color='red'>Задание 4.</font>
Разложить модуль $\;n\;$ на множители, если известно значение функции Эйлера.

In [7]:
n = 81533970462265904868952346508651285527924260726355932103005191827781540179809980478634968862010701112123363433327478138774759477160856150842111958981934522683839608508757908526048774711923204514049268169740286850383024571180686453826128804940189360350305379647199493425756784839375221754690420649798368968683
fi = 81533970462265904868952346508651285527924260726355932103005191827781540179809980478634968862010701112123363433327478138774759477160856150842111958981934504521046975582429330250778426468505794197178888272432634352526523304837798575506040635169000096959117771855371794488229073934582093291171352800845318041680

Применим формулы, полученные в предыдущей теореме.

Убедимся в правильности вычислений, используя Wolfram Mathematica. 

![wm](images/wm.png)

Для работы с длинными числами используется модуль decimal.

In [8]:
import decimal

decimal.getcontext().prec = 1_000  # количество знаков после запятой

A = decimal.Decimal(n + 1 - fi)
B = decimal.Decimal(A * A)
C = B - decimal.Decimal(4 * n)
D = decimal.Decimal(C).sqrt()

p = decimal.Decimal(A + D) / decimal.Decimal(2)
q = decimal.Decimal(A - D) / decimal.Decimal(2)

In [9]:
int(p * q) == n

True

## Эквивалентность знания секретной экспоненты и факторизации модуля $n$
**Теорема 3.**  
Задача вычисления секретной экспоненты $\;d\;$ полиномиально эквивалентна задаче разложения на множители модуля $\;n$ 
$$
  d\quad\Leftrightarrow\quad n=pq
$$

**Доказательство**  $\quad n=pq \quad\Rightarrow\quad d.$  
Если известно разложение $\;n=pq,\;$ то можно вычислить 
$$
\;\varphi(n)=(p-1)(q-1),\;
$$
а затем при помощи расширенного алгоритма Евклида найти 
$$
 d=e^{-1}\mod \varphi(n).
$$

$d\quad\Rightarrow\quad n=pq$.

Пусть известны $\;e,d.$

Так как $\;ed-1=z\varphi(n),\;$ то для $\;a\in\mathbb{Z}_n^*\;$ получим
$$
  a^{ed-1}\equiv (a^{\varphi(n)})^z\equiv 1^z\equiv 1\mod n.
$$

Так как $\;НОД(e,\varphi(n))=1,\;$ $\;НОД(d,\varphi(n))=1,\;$ то $\;e,d\;$ -- нечетные.

Представим
$$
ed-1=2^st,
$$
где $\;s,t\in\mathbb{N},\;$ $\;t\;$ -- нечетное.

Если найти такое $\tilde s\leq s\;$ и $\;a\in\mathbb{Z}_p^*,\;$ что
$$
v:=a^{2^{\tilde s}t}\equiv 1\mod n,\qquad u:=a^{2^{\tilde s-1}t}\not\equiv 1\mod n,
$$

то для $\;u\;$ справедливы сравнения 
$$
  u^2\equiv v\equiv 1\mod n\qquad \Rightarrow\qquad u^2-1\equiv (u-1)(u+1)\equiv 0\mod n.
$$
  
Если $\;u\neq-1,\;$ то $\;p:=\gcd(u-1,n),$  $q:=\gcd(u+1,n).$
 

## Алгоритм поиска $\;p,q\;$  по известной секретной экспоненте $\;d$
Input: $\;n, e, d\in\mathbb{N}.$  
   
Output: $\;p,q,\;$ что $\;n=p q.$
   
1. При помощи последовательного деления на 2 представляем $\;ed-1\;$ в виде $\;2^st,\;$ где $\;t\;$ -- нечетное.

2. Выбираем случайное $\;a\in\mathbb{N},\;$ $\;1<a<n-1.\;$

3. Если $\;\gcd(a,n)>1,\;$ то $\;p:=\gcd(a,n),\;$ $\;q:=\frac np,\;$ конец алгоритма.

4. Вычисляем $\;u:=a^t\mod n,\;$ $\;v:=u^2\mod n.$

5. Если $\;u=1,\;$ то переходим к шагу 2.

6. Пока $\;v\neq1\;$ вычисляем $\;u:=v,\;$ $\;v:=u^2\mod n.$

7. Если $\;u\equiv-1\mod n,\;$ то переходим к шагу 2, иначе выдаем результат
$$
\;p:=\gcd(u+1,n),\qquad\;q:=\gcd(u-1,n).
$$ 



## Оценка сложности алгоритма
-  На шаге 6 за один цикл выполняется не более $\;s\;$ $\;(s<2\log_2n)\;$ возведений в квадрат.

- При $\;n=p q\;$ уравнение 
$$
x^2\equiv 1\mod n
$$
имеет четыре корня, два из них $\;\pm1.$ 

- Мы хотим найти корни, отличные от $\;\pm1.$ 

- На 6 шаге мы находим $\;u\;$ -- один из корней данного уравнения, отличный от 1.    

- С вероятностью $\;\frac{2}{3}\;$ корень $\;u\;$ отличен от $\;-1\;$ и мы получим получим результат.


## Алгоритм Лас-Вегас
Алгоритм Лас-Вегас — вид вероятностного алгоритма.

Если у нас есть некий вероятностный алгоритм $A,$ который с определенной вероятностью дает верный результат, и существует возможность алгоритмически проверить результат алгоритма $A$ на корректность (скажем, с помощью алгоритма $K$), то можно выполнять алгоритм $A$ до тех пор, пока проверка не установит, что результат верен.

$$
Выполнить\; алгоритм\;\; A \;\;с\; результатом\;\; r\;\; до\; тех\; пор,
$$
$$
 \; пока\;\; K(r)\;\; не\; будет\; истиной.
$$


# 4. Пакеты в Python с реализацией RSA

## Модуль `crypto` пакета `SymPy`
Для генерации ключей криптосистемы RSA в данном модуле простые числа нужно генерировать самому

In [10]:
from sympy.crypto import crypto
from sympy.ntheory import randprime

b = 100
p = randprime(1<<b-1, 1<<b)
q = randprime(1<<b-1, 1<<b)
e = 7 
ke = crypto.rsa_public_key(p,q,e)
ke  # возвращает (n, e)

(614455700454306099640414072658602739557336986053803662542417, 7)

## Шифрование и дешифрование в `SymPy`
Шифруются числа из промежутка от 0 до n

In [11]:
message = 12
ciphertext = crypto.encipher_rsa(message, ke)
print(ciphertext)
type(ciphertext)

35831808


int

In [12]:
kd = crypto.rsa_private_key(p, q, e)
print(kd)
plaintext = crypto.encipher_rsa(ciphertext, kd)
print(message == plaintext)
plaintext

(614455700454306099640414072658602739557336986053803662542417, 351117543116746342651665184375439449432032492306059751516343)
True


12

## <font color='red'>Задание 5.</font>
По известному открытому ключу и шифротексту найти открытый текст.

In [13]:
e = 3
n = 12333277788503538360490919611266800078024649569685951268463814430930629204735581469679372448433345586153599652145159294063614796396444823879029701412108689560407872613358083071355514408041616230482676296185464003661906965152381896043173697776483174977213603631061284540519818689209796390301594178879738667603279453611728728974661852736510675375079653182192891218571020636895445524335114804592353717984205231217785408769204077045632362456543752496623299520523221151683581539402696249769884091094137519125308510737426376762698303050958197835205893061971688021772454949648476014529948682797820777748360390484713587284601
ciphertext = 125  # зашифрованный текст

In [14]:
key_e = (n, e)
message = 0

while True:
    ciphertext_new = crypto.encipher_rsa(message, key_e)
    
    if ciphertext == ciphertext_new:
        break
    else:
        message += 1
        
print('Открытый текст: ', message)

Открытый текст:  5


## Пакет RSA

In [15]:
!pip install RSA

In [15]:
import rsa
(bob_pub, bob_priv) = rsa.newkeys(512)  # генерирует ключи по количеству бит
print('bob_pub = ', bob_pub)  # (n, e)
print('bob_priv = ', bob_priv)  # (n, e, d, p, q)

bob_pub =  PublicKey(10528509047744953485684851684142049999804385856279173507664180697978440234793764723905162731285308836595683623276876298577971198281626731572587676728571801, 65537)
bob_priv =  PrivateKey(10528509047744953485684851684142049999804385856279173507664180697978440234793764723905162731285308836595683623276876298577971198281626731572587676728571801, 65537, 8747062890147813103873671429851325791375088897915504816291549973961733273040478064252779405036055903496666802093170022277868475616062428235390555439741889, 7140704722119707400255400431099159760350297833907187367826690240264090307328946959, 1474435571482303429235484849394766759006210911588313596817188398788135639)


In [16]:
message = 'hello Bob!'.encode('utf8')
ciphertext = rsa.encrypt(message, bob_pub)
print(ciphertext)

b'T!N\x8d`\xfb\xfd9\x0c\xbf\xc1ta\xb0R\xa2\xc9F\xc8\x1b\xc0\xa9D\xb0\x1b\x19\x90\x06\x06\x86\x8b\xbao\xaa\x1b\xc6\x8ao\x80,}1\xd4mM\xec\\^\xf4\xfe\xbf\xd83\t#\xf5\xd4\xfc@W|q\xa9c'


In [17]:
plaintext = rsa.decrypt(ciphertext, bob_priv)  # для расшифровки нужен объект класса rsa.PrivateKey
print(plaintext.decode('utf8'))
message == plaintext

hello Bob!


True

## <font color='red'>Задание 6.</font>
По известному модулю $\;n,\;$ открытой экспоненте $\;e\;$ и секретной экспоненте $\;d\;$ найти разложение модуля $\;n\;$ на множители. Сравнить количество бит, из которых состоят числа $\;p,\;$ $\;q\;$ в криптосистеме RSA из пакета RSA. 

In [18]:
pk = rsa.PublicKey(7161119666703588511485958969935251388960044601326459148749567084396681512566112122272864126455327982625175589758995636362943052020959413616257193856359801, 65537)
d = 2550323222315054943895544232392217639170658421883204243889938443166738580432008124211120675369247637004694114557452402924433022731484378117056379857281473

In [19]:
def private_key_search_p_q(key_e: tuple, d: int) -> tuple:
    
    """ search p and q using a private key """
    
    n, e = key_e
    s, t = st(e * d)  # не вычитаем 1, так как это уже сделано в теле функции st
    
    while True:
        a = randint(1 + 1, n - 2)

        if math.gcd(a, n) > 1:
            p = math.gcd(a, n)
            q = n // p

            return (p, q)
        else:
            u = pow(a, t, n)
            v = pow(u, 2, n)

            if u == 1:
                continue
            else:
                while v != 1:
                    u = v
                    v = pow(u, 2, n)
                
                if (u + 1) % n == 0:
                    continue
                else:
                    p = math.gcd(u + 1, n)
                    q = math.gcd(u - 1, n)
                    
                    return (p, q)

In [20]:
key_e = (pk.n, pk.e)

p, q = private_key_search_p_q(key_e, d)

p * q == pk.n

True

In [21]:
len(bin(p)[2:])

272

In [22]:
len(bin(q)[2:])

240

Из вышеприведённой информации видно, что количество бит из которых состоят числа $p, q$ в криптосистеме `RSA` из пакета `RSA` соответственно равны $272$ и $240$, следовательно, в общем случае при передаче в функцию `rsa.newkeys` b бит, получим, что $p$ будет состоять из $(17/32)b = 0.53125\;b$ бит, а $q$ $-$ из $(15/32)b = 0.46875\;b$ бит соответсвенно.

## Пакет PyCrypto
Пакет `PyCrypto` на мою старую операционную систему Windows7 не установился.

In [23]:
!pip install PyCrypto

  Using cached pycrypto-2.6.1.tar.gz (446 kB)
  Running setup.py clean for PyCrypto
Failed to build PyCrypto
    Running setup.py install for PyCrypto: started
    Running setup.py install for PyCrypto: finished with status 'error'


  ERROR: Command errored out with exit status 1:
   command: 'D:\Anaconda\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\37529\\AppData\\Local\\Temp\\pip-install-njpxgoe3\\pycrypto_05d71792e62e4b41972292045987661c\\setup.py'"'"'; __file__='"'"'C:\\Users\\37529\\AppData\\Local\\Temp\\pip-install-njpxgoe3\\pycrypto_05d71792e62e4b41972292045987661c\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\37529\AppData\Local\Temp\pip-wheel-w9zlwjck'
       cwd: C:\Users\37529\AppData\Local\Temp\pip-install-njpxgoe3\pycrypto_05d71792e62e4b41972292045987661c\
  Complete output (183 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd64-3.9
  creating bui

    D:\Windows Kits\10\include\10.0.19041.0\ucrt\inttypes.h(51): error C2146: синтаксическая ошибка: отсутствие ")" перед идентификатором "_Numerator"
    D:\Windows Kits\10\include\10.0.19041.0\ucrt\inttypes.h(51): error C2061: синтаксическая ошибка: идентификатор "_Numerator"
    D:\Windows Kits\10\include\10.0.19041.0\ucrt\inttypes.h(51): error C2059: синтаксическая ошибка: ;
    D:\Windows Kits\10\include\10.0.19041.0\ucrt\inttypes.h(51): error C2059: синтаксическая ошибка: ,
    D:\Windows Kits\10\include\10.0.19041.0\ucrt\inttypes.h(53): error C2059: синтаксическая ошибка: )
    D:\Windows Kits\10\include\10.0.19041.0\ucrt\inttypes.h(55): error C2143: синтаксическая ошибка: отсутствие "{" перед "__cdecl"
    D:\Windows Kits\10\include\10.0.19041.0\ucrt\inttypes.h(61): error C2143: синтаксическая ошибка: отсутствие "{" перед "__cdecl"
    D:\Windows Kits\10\include\10.0.19041.0\ucrt\inttypes.h(68): error C2143: синтаксическая ошибка: отсутствие "{" перед "__cdecl"
    D:\Windows K

# Пакет cryptography

### Генерируем секретный ключ:

In [24]:
from cryptography.hazmat.primitives.asymmetric import rsa
private_key = rsa.generate_private_key(public_exponent=65537,key_size=2048)

private_numbers = private_key.private_numbers()
print(len(bin(private_numbers.p))-2)
print(len(bin(private_numbers.q))-2)

1024
1024


### Из секретного ключа находим открытый ключ

In [25]:
public_key = private_key.public_key()
public_key.public_numbers()

<RSAPublicNumbers(e=65537, n=23594776944540285883465924215270456910889592691068761320304520846830788916032431143745358728258111429040084242205190144548985188177465325739245386978821640001530471345397848094150932294236102721126025625822626015832434426042925985481268962490283703221960667942947443481002583307581708180653067658724900942207787799099436216857422295464435303987671293823084490372111775906229592829064792515409998524991022234888005920685825202354245837746040938971715743252121905401359339915439966148640991350314966943457591977955892747178174689689450362708680841101490951313434971095540175432047476479216105419869964926661701597314691)>

## Шифрование
Шифруем сообщение, используя алгоритмы дополнения сообщения `padding.OAEP`. Более подробно об алгоритмах дополнения сообщения можно будет узнать на следующей лекции

In [26]:
from cryptography.hazmat.primitives.asymmetric import padding
from cryptography.hazmat.primitives import hashes

message = b"Hello  KM!"
ciphertext = public_key.encrypt(
    message,
    padding.OAEP(
        mgf=padding.MGF1(algorithm=hashes.SHA256()),
        algorithm=hashes.SHA256(),
        label=None
    )
)
ciphertext.hex()

'2259409d0ebab6b33acf11d282a002e5fe5e2c23345551433352364dab88261ee8a733de86d240b9a21d82bdef0b4912aa878ee7168a762af8041a8f05674db13270d90f15f3e47b5329d8bc79efe980fcb4aa058fe03f321f857d6aa0fda8310ec7dd741ea7f2c181fa3a8dac958d60f7e2f8a738656008b45095606aeac5600aaf6a1ca2f12d3024382d64f48f78c399c6c6b93b9c570f303b49710679a4ac2458660e684bb6b7e31809585e743eb583cc2c2ae78de0c703a62c4a1237221186edee9adbda439dfee8548d665314c28700c1b56f815441bd66722d077fbb66958353a864a0f2175228f0d4e970cf4f132d8e2c84d823e4e69ffe592a3493a4'

## Дешифрование

In [27]:
plaintext = private_key.decrypt(
    ciphertext,
    padding.OAEP(
        mgf=padding.MGF1(algorithm=hashes.SHA256()),
        algorithm=hashes.SHA256(),
        label=None
    )
)
print(plaintext)
plaintext == message

b'Hello  KM!'


True

## <font color='red'>Задание 7.</font>
Если модуль `n` имеет специальный вид, то криптосистему RSA можно взломать.
По известному модулю `n`, открытой экспоненте `e` и шифротексту `ciphertext` найти открытый текст. Какой дефект числа `n` повлиял на успешность взлома?

In [28]:
e = 65537 
n = 132935059133128588099304661920243934739765436275380706551932167743169033487495503272227432220579458014156688684426929893492508715387507321932857109769559317685058136392287296762655125670378191567317846796377276529558623179825406400952782093978732505373117024648318275223210130826397937612067898131709502900576386539177352833994857536083365047052365456565284977278734384796276824486670451880158184829081497950883
ciphertext = 33402684541071019355683217046735370114635288822383014565136096069237202317507833906996968322171616228317453040371636219641462611973676323964271013324667958657860405122888987565765804160949370961090578985257116551787982001404003947199732093633057535739007675152772802674565570756158429484673327638366744917261367473853248255643173343336198861934268542799366124227922100857141014227177919922090578492025680513581

In [29]:
key_e = (n, e)
message = 0

while True:
    ciphertext_new = crypto.encipher_rsa(message, key_e)
    
    if ciphertext == ciphertext_new:
        break
    else:
        message += 1
        
print('Открытый текст: ', message)

Открытый текст:  257


Про уязвимости криптосистемы RSA можно прочитать по <a align="https://habr.com/ru/company/virgilsecurity/blog/459370/">ссылке<a>.